src/aof.c

```diff
@@ -40,191 +40,603 @@
+ *   server.aof_filename.seq.aof
+ */
+const char* getNewIncrAofName(aofManifest *am) {
+    aofInfo *ai = aofInfoCreate();
+    ai->file_type = AOF_FILE_TYPE_INCR;
+    ai->file_name = sdscatprintf(sdsempty(),"%s.%lld%s",server.aof_filename, 
+                        ++am->curr_incr_aof_seq,INCR_AOF_SUFFIX);
+    ai->file_seq = am->curr_incr_aof_seq;
+    listAddNodeTail(am->incr_aof_list,ai);
+
+    am->dirty = 1;
+    return ai->file_name;
+}
+
+/* Get the last INCR AOF name or create a new one. */
+const char *getLastIncrAofName(aofManifest *am) {
+    serverAssert(am != NULL);
+
+    /* incr_aof_list is empty, just create a new one. */
+    if (listLength(am->incr_aof_list) == 0) {
+        return getNewIncrAofName(am);
+    }
+
+    /* Or return the last one. */
+    listNode *lastnode = listIndex(am->incr_aof_list,-1);
+    aofInfo *ai = listNodeValue(lastnode);
+    return ai->file_name;
+}
+
+/* Called in `backgroundRewriteDoneHandler`. when AOFRW success, This
+ * function will change the AOF file type in `incr_aof_list` from AOF_FILE_TYPE_INCR 
+ * to AOF_FILE_TYPE_HIST, and move them to the `history_aof_list`.
+ */
+void markRewrittenIncrAofAsHistory(aofManifest *am) {
+    serverAssert(am != NULL);
+
+    if (listLength(am->incr_aof_list) == 0) {
+        return;
+    }
+    
     listNode *ln;
-    aofrwblock *block;
-    ssize_t nwritten;
-    mstime_t latency;
-    UNUSED(el);
-    UNUSED(fd);
-    UNUSED(privdata);
-    UNUSED(mask);
+    listIter li;
 
-    latencyStartMonitor(latency);
-    while(1) {
-        ln = listFirst(server.aof_rewrite_buf_blocks);
-        block = ln ? ln->value : NULL;
-        if (server.aof_stop_sending_diff || !block) {
-            aeDeleteFileEvent(server.el,server.aof_pipe_write_data_to_child,
-                              AE_WRITABLE);
-            break;
-        }
-        if (block->used != block->pos) {
-            nwritten = write(server.aof_pipe_write_data_to_child,
-                             block->buf+block->pos,block->used-block->pos);
-            if (nwritten <= 0) break;
-            block->pos += nwritten;
-        }
-        if (block->used == block->pos) listDelNode(server.aof_rewrite_buf_blocks,ln);
+    listRewindTail(am->incr_aof_list,&li);
+
+    /* server.aof_fd != -1 means that AOF is open, then we must skip the last 
+     * AOF, because this one is our currently writing. */
+    if (server.aof_fd != -1) {
+        ln = listNext(&li);
+        serverAssert(ln != NULL);
     }
-    latencyEndMonitor(latency);
-    latencyAddSampleIfNeeded("aof-rewrite-write-data-to-child",latency);
+
+    /* move aofInfo from incr_aof_list to history_aof_list. */
+    while ((ln = listNext(&li)) != NULL) {
+        aofInfo *ai = (aofInfo*)ln->value;
+        serverAssert(ai->file_type == AOF_FILE_TYPE_INCR);
+
+        aofInfo *hai = aofInfoDup(ai);
+        hai->file_type = AOF_FILE_TYPE_HIST;
+        listAddNodeTail(am->history_aof_list,hai);
```

Comment:
Here I want to ensure the order they were before.

Reply:
yes, i'm arguing that you're reversing the order.
maybe i'm missing something... 
since you're iterating the old list from tail to head, if you're inserting to the tail of the new list, it'll reverse the order.

Reply:
Yes you are right, i am wrong. fixed ：)